### PR TITLE
Exercise ticket with different collateral

### PR DESCRIPTION
**Pull request by kirilaa**
- thales-dao/contracts is updated
- multicollateral.priceFeed() is used since `priceFeed()` is added into IMultiCollateralOnOffRamp
- receive() payable added for offramping into ETH
- `_exercise(address _ticket)` is extended to -> `_exerciseTicket(address _ticket, address _exerciseCollateral, bool _inEth)`
- Ticket.exercise() returns finalPayout

### CODE REVIEW

File: contracts/Overtime/SportsAMMV2.sol [hunk 1]
```diff
@@ -381,7 +385,19 @@ contract SportsAMMV2 is Initializable, ProxyOwned, ProxyPausable, ProxyReentranc
...
+    /// @notice exercise specific ticket to an off ramp collateral
+    /// @param _ticket ticket address
+    /// @param _exerciseCollateral collateral address to off ramp to
+    /// @param _inEth offramp with ETH
+    function exerciseTicketOffRamp(
+        address _ticket,
+        address _exerciseCollateral,
+        bool _inEth
+    ) external nonReentrant notPaused onlyKnownTickets(_ticket) {
+        _exerciseTicket(_ticket, _exerciseCollateral, _inEth);
```
Comment (vladanthales) [inline contracts/Overtime/SportsAMMV2.sol:400]: Missing check that only ticket owner can call this method

File: contracts/Overtime/SportsAMMV2.sol [hunk 2]
```diff
@@ -757,11 +773,39 @@ contract SportsAMMV2 is Initializable, ProxyOwned, ProxyPausable, ProxyReentranc
         );
     }
 
-    function _exerciseTicket(address _ticket) internal {
+    function _exerciseTicket(address _ticket, address _exerciseCollateral, bool _inEth) internal {
         Ticket ticket = Ticket(_ticket);
-        ticket.exercise();
-        IERC20 ticketCollateral = ticket.collateral();
+        (uint userWinningAmount, address ticketOwner, IERC20 ticketCollateral) = ticket.exercise(_exerciseCollateral);
```
Comment (vladanthales) [inline contracts/Overtime/SportsAMMV2.sol:778]: ticketOwner and ticketCollateral can be called directly from Ticket

File: contracts/Overtime/SportsAMMV2.sol [hunk 3]
```diff
@@ -757,11 +773,39 @@ contract SportsAMMV2 is Initializable, ProxyOwned, ProxyPausable, ProxyReentranc
...
+                // this method is used as a suggestion from here:
+                // https://solidity-by-example.org/sending-ether/
+                (bool sent, ) = payable(ticketOwner).call{value: offramped}("");
+                require(sent, "Failed to send Ether");
+            } else {
+                // Multi-collateral contract needs changes, for example to swap from ARB -> USDT
+                // Approval code can be deleted for size/gas savings.
+                if (
+                    IERC20(_exerciseCollateral).allowance(address(this), address(multiCollateralOnOffRamp)) <
+                    userWinningAmount
```
Comment (vladanthales) [inline contracts/Overtime/SportsAMMV2.sol:796]: I think allowance should be checked on `ticketCollateral` not `_exerciseCollateral`.

File: contracts/Overtime/SportsAMMV2.sol [hunk 4]
```diff
@@ -757,11 +773,39 @@ contract SportsAMMV2 is Initializable, ProxyOwned, ProxyPausable, ProxyReentranc
...
-        IERC20 ticketCollateral = ticket.collateral();
+        (uint userWinningAmount, address ticketOwner, IERC20 ticketCollateral) = ticket.exercise(_exerciseCollateral);
         uint amount = ticketCollateral.balanceOf(address(this));
+        if (userWinningAmount > 0 && _exerciseCollateral != address(0) && _exerciseCollateral != address(ticketCollateral)) {
+            // Off ramp methods
+            // Work in progress:
+            uint offramped;
+            if (_inEth) {
+                // Multi-collateral contract needs changes, for example to swap from ARB -> ETH
+                offramped = multiCollateralOnOffRamp.offrampFromIntoEth(address(ticketCollateral), userWinningAmount);
```
Comment (vladanthales) [inline contracts/Overtime/SportsAMMV2.sol:786]: Allowance should be checked in this case also.

**Review by vladanthales — CHANGES_REQUESTED** [review]

File: contracts/Overtime/SportsAMMV2.sol (hunk 3 above)
Comment (danijelthales) [reply contracts/Overtime/SportsAMMV2.sol:796]: Its always default collateral, and MultiCollateral contract has infinite approval, so not needed.

File: contracts/Overtime/SportsAMMV2.sol (hunk 1 above)
Comment (kirilaa) [reply contracts/Overtime/SportsAMMV2.sol:400]: Ok, yea makes sense only the ticketOwner to call offramp
and default exercise by anyone

File: contracts/Overtime/SportsAMMV2.sol (hunk 2 above)
Comment (kirilaa) [reply contracts/Overtime/SportsAMMV2.sol:778]: saves gas by adding 3 external calls into a single call

File: contracts/Overtime/SportsAMMV2.sol (hunk 3 above)
Comment (kirilaa) [reply contracts/Overtime/SportsAMMV2.sol:796]: will modify to defaultCollateral

File: contracts/Overtime/SportsAMMV2.sol [hunk 5]
```diff
@@ -757,11 +773,39 @@ contract SportsAMMV2 is Initializable, ProxyOwned, ProxyPausable, ProxyReentranc
...
-        ticket.exercise();
-        IERC20 ticketCollateral = ticket.collateral();
+        (uint userWinningAmount, address ticketOwner, IERC20 ticketCollateral) = ticket.exercise(_exerciseCollateral);
         uint amount = ticketCollateral.balanceOf(address(this));
+        if (userWinningAmount > 0 && _exerciseCollateral != address(0) && _exerciseCollateral != address(ticketCollateral)) {
+            // Off ramp methods
+            // Work in progress:
+            uint offramped;
+            if (_inEth) {
+                // Multi-collateral contract needs changes, for example to swap from ARB -> ETH
```
[the remark this answers is not in the document]
Comment (kirilaa) [reply contracts/Overtime/SportsAMMV2.sol:785]: ok, will change only to defaultCollateral

**Review by vladanthales — DISMISSED** [review]

**Review by vladanthales — APPROVED** [review]

**Review by danijelthales — DISMISSED** [review]

**Review by vladanthales — DISMISSED** [review]

**Review by vladanthales — APPROVED** [review]

**Review by danijelthales — APPROVED** [review]

**Pull request merged**